### DISCUSSION TITLE
Nested form error with `touched` and `dirty`

### DISCUSSION THREAD

**Discussion by iKrushYou** [react-hook-form/react-hook-form#5349] [Q&A] [answered]
**Describe the bug**
This bug seems to arise when trying to track the `touched` and `dirty` states of custom `Controller` components that have nested values. In this example, I have set up a few scenarios with issues i have run into. There are 4 fields in the UI and I will refer to them in order top to bottom.

Field 1 and 2 have different names

Field 1 - touch and dirty works as expected. If you select the field and then leave, the `Touched` value in the dev tools will update. If you type into the field, `Dirty` will update as well
Field 2 - `Touched` does not work, nor does `Dirty`
---
Field 3 and 4 have the same name but one is nested: `exampleRequried2` and `nested.exampleRequred`

Field 3 - if you touch this field, it will update `Touched` for both field 3 and field 4. Same with `Dirty` even though the value won't update.
Field 4 - same as Field 2 except now the state is affected by field 3

**Codesandbox link (Required)**
https://codesandbox.io/s/react-hook-form-issue-with-controlled-nested-fields-urm00?file=/src/index.tsx

**Comment by bluebill1049** [comment, marked as answer]
<img width="588" alt="Screen Shot 2021-05-27 at 9 35 07 am" src="https://user-images.githubusercontent.com/10513364/119744190-d457db00-bece-11eb-9575-17dcb70ef140.png">

where are the touchedFields, and dirtyFields subscriptions?

have a read on this section: https://react-hook-form.com/api/useform/formstate

**Reply by iKrushYou** [reply]
I'm observing through the use of DevTools

**Reply by bluebill1049** [reply]
i see. it's a devtool bug.

**Reply by bluebill1049** [reply]
https://github.com/react-hook-form/devtools/issues/84

**Reply by iKrushYou** [reply]
I am definitely seeing issues outside of devtools though when using nested fields

**Reply by iKrushYou** [reply]
I'll see if I can isolate it

**Reply by bluebill1049** [reply]
> I am definitely seeing issues outside of devtools though when using nested fields

if you can reproduce that outside of devtool, we can re-open the issue.

**Reply by iKrushYou** [reply]
okay I think I reproduced it, it's an issue with the yup resolver when you have a nested field that has a `when` condition to another field

**Reply by iKrushYou** [reply]
https://codesandbox.io/s/react-hook-form-issue-with-yup-when-validation-nested-fields-o3pxc?file=/src/index.tsx

**Reply by bluebill1049** [reply]
What's the steps to reproduce? Is it the same issue with dirty and touched

**Reply by iKrushYou** [reply]
It's an issue with validation. Hopefully the input labels are hpful enough to figure it out 

If you validate the last two inputs before / after the conditional field 1 has `test` as a value

**Reply by bluebill1049** [reply]
This is dependant validation, right? you will need to use `trigger` if that's the case, RHF focuses single input at a time.

**Reply by iKrushYou** [reply]
This is yup validation dependent on another input

It works as expected when the inputs aren't nested

**Reply by bluebill1049** [reply]
I would investigate your schema. https://codesandbox.io/s/react-hook-form-issue-with-yup-when-validation-nested-fields-forked-2krkd?file=/src/index.tsx

<img width="1745" alt="Screen Shot 2021-05-27 at 12 38 05 pm" src="https://user-images.githubusercontent.com/10513364/119757517-64eee500-bee8-11eb-8fa5-cd329b43c58c.png">

**Reply by iKrushYou** [reply]
Thank you @bluebill1049 
I appreciate the responses so far. Would you mind elaborating? I have combed over your image, forked CSB, and my own code and I still cannot figure out what could be the issue. The console logs you have don't show anything different than we're seeing in the UI. Perhaps the variables are poorly named but what I would expect is:
just like `exampleRequired2` ends up being a required field because it is dependent on `exampleRequired` being `test`.. I would expect `nested.exampleRequired2` to also have an error which it does not. However, `nested.exampleRequiredNested` does show an error as it was not dependent on another field in yup

**Reply by bluebill1049** [reply]
did you see the output for this part: 

```
try {
  schmea.validateSync(data, { abortEarly: false });
} catch (e) {
  console.log("error", e.errors);
}
```

hook form does no magic by sending data and gets validate by yup, by all means, I am not a yup expert, if the data supplied is correct and the validation is what your schema is defined, I think everything is work for the lib. @jorisre maybe you can give a hand here?

**Reply by iKrushYou** [reply]
@bluebill1049 ah.. thank you. I am sorry I didn't see it sooner that the issue is actually with yup and the fact that yup doesn't allow nested validation against parents :| 

Sorry to bother! But hey at least I found the issue with dev tools ¯\_(ツ)_/¯

**Reply by bluebill1049** [reply]
No worries at all, please don't hesitate to report issues. thanks.

**Comment by iKrushYou** [comment]
It was quite hacky but I was able to solve my issue with yup

```javascript
const schmea = yup.object().shape({
  field1: yup.string().required(),
  field2: yup.string().when("field1", {
    is: "test",
    then: yup.string().required()
  }),
  nested: yup.object().shape({
    field3: yup.string().required(),
    field4: yup
      .string()
      .test("field4-test", "Field Required", (value, context) => {
        // @ts-ignore
        if (context.from[1].value.field1 === "test") return !!value;
        return true;
      }),
    field5: yup.string().when("field3", {
      is: "test",
      then: yup.string().required()
    })
  })
});
```

and specifically the test function
```javascript
    field4: yup
      .string()
      .test(
        "field4-test",
        "Field Required",
        (value, context) => {
          // @ts-ignore
          if (context.from[1].value.field1 === "test") return !!value;
          return true;
        }
      ),
```

https://github.com/jquense/yup/pull/556#issuecomment-641988687

There are not TS types at the moment so it's very inelegant

**Reply by iKrushYou** [reply]
Or you can cast it
```javascript
const contextGeneric = (context as unknown) as { from: { value: Record<string, unknown> }[] };
```
